test/passes/O3_low-memory-unused_metrics.txt

```diff
@@ -1084,10 +1084,10 @@ total
                   (loop $label$28
                    (if
                     (i32.eq
+                     (local.get $3)
```

Comment:
It looks like this change actually reduced the amount of canonicalization that happens. If the result of a pattern is not canonical, it used be be canonicalized for further optimizations, but this PR prevents that. Is that concerning?

Reply:
Yes, it reduce canonical in some rare cases. But all this changes not really important and can't interfere with matching.

Reply:
Is the performance benefit from this change significant? If not, it's probably best not to do this since it makes canonicalization less robust.

Reply:
benchmark for unoptimized `as.wasm` (~1.1 MB) with options:
```bash
BINARYEN_PASS_DEBUG=1 ./wasm-opt as.wasm -o out.wasm --optimize-instructions --enable-mutable-globals --converge
```

### master:

passes took 0.0387452 seconds.
passes took 0.0300471 seconds.

### this pr:  

passes took 0.033229 seconds. 
passes took 0.0285829 seconds.

Reply:
16% is more than I expected here, but it is 16% on one of the fastest passes we have. So I don't think it's very significant.

I think it would still be worth doing if there wasn't the downside, but I think the downside is a problem.

To avoid the downside, in theory we could, as this PR does, not canonicalize in the inner loop, but then we should also have an assertion in there in debug builds that the intermediate outputs are actually canonicalized, so we never miss anything. And we'd need to do the work to fix all those things. And, we'd need to test more with debug builds to catch errors here.

Overall I think this is an interesting idea, but I suspect it's not worth the effort. If this pass ever gets slow enough that we need to look at speeding it up, there may be safer/easier things to do.

Reply:
Alright) I'll revert back canonization back into loop, but the rest changes which not affect to canonization keep here. WDYT?